boolean/test_boolean.py

```diff
@@ -1165,6 +1166,32 @@ def test_objects_return_set_of_unique_Symbol_objs(self):
         exp = alg.parse('a and b or a and c')
         assert set(['a', 'b', 'c']) == exp.objects
 
+    def test_normalize_blowup(self):
+        # Real-world example of a complex expression with simple CNF/DNF form.
+        # Note this is a more reduced, milder version of the problem, for rapid
+        # testing.
+        formula = """
+        a & (
+            (b & c & d & e & f & g)
+            | (c & f & g & h & i & j)
+            | (c & d & f & g & i & l & o & u)
+            | (c & e & f & g & i & p & y & ~v)
+            | (c & f & g & i & j & z & ~(c & f & g & i & j & k))
+            | (c & f & g & t & ~(b & c & d & e & f & g))
+            | (c & f & g & ~t & ~(b & c & d & e & f & g))
+        )
+        """
+        algebra = BooleanAlgebra()
+        expr = algebra.parse(formula)
+        t0 = time.time()
+        cnf = algebra.cnf(expr)
+        t1 = time.time()
+
+        assert str(cnf) == "a&c&f&g"
+        # Locally, this test takes 0.4s, previously it was 500s.
+        # We allow 30s because of the wide range of possible CPUs.
+        assert t1 - t0 < 30, "Normalizing took too long"
```

Comment:
We shouldn't use runtime in an assertion as a general rule... it's brittle as you already indicated in the comment. You may want to explore using mocks and counting function calls. `expr.simplify` looks like a good candidate.

Reply:
@prodhype @olliemath should I go ahead and merge though?

Reply:
@pombredanne It should be rewritten so that it doesn't depend on runtime before being merged.

Reply:
@prodhype there you go. Thanks!

Reply:
See https://github.com/bastikr/boolean.py/pull/107/commits/5f93c8b17440348ffe024c8498e48c5bfa5ae0be

Reply:
@pombredanne sorry it took me a while to get back to you, I've been a bit too busy for much open source recently. Thanks for improving the test though!